coilsnake/modules/eb/MusicModule.py

```diff
@@ -0,0 +1,293 @@
+                        for song in pack_obj.songs:
+                            self.songs[song.song_number] = song
+                except InvalidUserDataError as e:
+                    raise InvalidUserDataError("Error reading pack ${:02X} as {}: {}".format(pack_num, pack_type.__name__, e.message))
+            else:
+                log.warn("Music pack ${:02X} contains no data.".format(pack_num))
+                pack_obj = mp.EmptyPack(pack_num)
+            self.packs.append(pack_obj)
+        return
+
+    def write_to_rom(self, rom):
+        # Apply Gas Station instrument pack patch
+        # In the vanilla game, there is a bug in the INITIALIZE_SPC700 function where it loads the
+        # engine pack by looking at the sequence pack of song 00, but then stores the loaded pack
+        # value into the secondary instrument pack variable. This means when we play the
+        # Gas Station song, it will load the whole engine again, which includes some instrument
+        # data. This instrument data will be loaded after the instrument packs, so this can result
+        # in some corrupted instruments when changing the instrument packs used by Gas Station.
+        self.get_patch(rom).apply(rom)
+        # Prepare packs for writing out by converting them to parts
+        for pack in self.packs:
+            pack.save_to_parts()
+        # Build song address table and song pack table
+        for song_num, song in self.songs.items():
+            song_ind = song_num - 1
+            self.song_address_table[song_ind] = song.get_song_aram_address()
+            self.song_pack_table[song_ind] = song.get_song_packs()
+        # Write song address table into engine pack
+        if not isinstance(self.packs[1], mp.EngineMusicPack):
+            raise InvalidUserDataError("Expected pack 1 to be the music engine pack, instead it is of type {}".format(
+                type(self.packs[1]).__name__))
+        song_address_table_data = Block(self.song_address_table.size)
+        self.song_address_table.to_block(song_address_table_data, 0)
+        self.packs[0x01].set_song_address_table_data(song_address_table_data)
+        # Write out packs
+        for i, pack in enumerate(self.packs):
+            data = pack.get_pack_binary_data()
+            pack_offset = rom.allocate(data=data)
+            self.pack_pointer_table[i] = [to_snes_address(pack_offset)]
+        # Write out pack pointer table
+        self.pack_pointer_table.to_block(block=rom, offset=from_snes_address(self.PACK_POINTER_TABLE_ROM_ADDR))
+        # Build song pack table
+        self.song_pack_table.to_block(block=rom, offset=from_snes_address(self.SONG_PACK_TABLE_ROM_ADDR))
+        return
+
+    def get_patch(self, rom):
+        ips = IpsPatch()
+        ips.load(get_ips_filename(rom.type, 'gas_station_pack_fix'), 0)
+        return ips
+
+    def upgrade_project(self, old_version, new_version, rom, resource_open_r, resource_open_w, resource_delete):
```

Comment:
If there is future potential for changes to this module, it might be better to use the `UsedRangeModule`-style template for this method